File: templates/problems/15_greedy/01_max_subarray/python/solution.py
def solve_max_sub_array(nums: list[int]) -> int:
    """Kadane's algorithm. O(n) time, O(1) space."""
    max_sum = nums[0]
    cur_sum = nums[0]
    for i in range(1, len(nums)):
        if cur_sum < 0:
            cur_sum = nums[i]
        else:
            cur_sum += nums[i]
        if cur_sum > max_sum:
            max_sum = cur_sum
    return max_sum

File: templates/problems/15_greedy/01_max_subarray/typescript/solution.ts
/**
 * Kadane's algorithm. O(n) time, O(1) space.
 */
export function solveMaxSubArray(nums: number[]): number {
    let maxSum = nums[0];
    let curSum = nums[0];
    for (let i = 1; i < nums.length; i++) {
        if (curSum < 0) {
            curSum = nums[i];
        } else {
            curSum += nums[i];
        }
        if (curSum > maxSum) {
            maxSum = curSum;
        }
    }
    return maxSum;
}
